lib/id3c/cli/command/etl/fhir.py

```diff
@@ -0,0 +1,683 @@
+        scale = 'address'
+        address_identifier = identifier(location, f"{INTERNAL_SYSTEM}/location/{scale}")
+
+        if not address_identifier:
+            return None
+
+        return upsert_location(db,
+            scale       = scale,
+            identifier  = address_identifier,
+            hierarchy   = tract_hierarchy)
+
+
+    code = location_code(location)
+    relation = location_relation(code)
+
+    try:
+        parent_location = location.partOf.resolved(Location)
+        tract_hierarchy = get_tract_hierarchy(parent_location)  # TODO do we only want parent hierarchy?
+
+    except AttributeError:
+        LOG.info(f"No parent location found for Location with code «{code}», " + \
+            f"relation «{relation}».")
+        tract_hierarchy = None
+
+    address = process_address(location, tract_hierarchy)
+
+    if not (tract_hierarchy or address):
+        LOG.warning(f"No tract or address location available for «{relation}»")
+        return
+
+    upsert_encounter_location(db,
+        encounter_id    = encounter_id,
+        relation        = relation,
+        location_id     = address.id)
+
+
+def location_code(location: Location) -> str:
+    """
+    Given a *Location* Resource, returns the matching system code for our
+    internal, location-relation system.
+    """
+    codes: List[str] = []
+    codes += list(map(lambda t: matching_system_code(t, LOCATION_RELATION_SYSTEM), location.type))
+
+    if not codes:
+        return None
+
+    unique_codes = list(set(codes))
+    if len(unique_codes) > 1:
+        raise Exception(f"Expected only one Location code. Instead received {unique_codes}.")
+
+    return unique_codes[0]
+
+
+def process_sample(db: DatabaseSession, barcode: str) -> Any:
+    """ Given a *barcode*, returns its matching sample from ID3C. """
+    sample = find_sample(db, barcode)
+
+    if not sample:
+        raise SampleNotFoundError(f"No sample with «{barcode}» found.")
+
+    return sample
+
+
+def process_presence_absence_tests(db: DatabaseSession, report: DiagnosticReport,
+    sample_id: int, barcode: str):
+    """
+    Given a *report* containing presence-absence test results, upserts them to
+    ID3C, attaching a sample and target ID.
+    """
+    for result in report.result:
+        observation = result.resolved(Observation)
+
+        target_identifier = matching_system_code(observation.code, TARGET_SYSTEM)
+        # Most of the time we expect to see existing targets so a
+        # select-first approach makes the most sense to avoid useless
+        # updates.
+        target = find_or_create_target(db,
+            identifier  = target_identifier,
+            control     = False)
+
+        upsert_presence_absence(db,
+            identifier = f'{barcode}/{target_identifier}/{observation.device.identifier.value}',
+            sample_id = sample_id,
+            target_id = target.id,
+            present = observation.valueBoolean,
```

Comment:
Just flagging that we need to be able to ingest both `valueBoolean` and `valueCodeableConcept` per [slack convo](https://seattle-flu-study.slack.com/archives/CPE5K1848/p1576523853002400)